File: packages/dfos-protocol/src/chain/revocation.ts
/*

  REVOCATION

  Signed revocation artifact for DFOS credentials. Gossiped across the relay
  network like beacons. Permanent — no un-revoke, issue a new credential
  instead.

  Only the credential's issuer DID can revoke it. Relay maintains a revocation
  set per issuer and checks it during credential verification.

*/

import { createJws, dagCborCanonicalEncode, decodeJwsUnsafe, verifyJws } from '../crypto';
import { RevocationPayload } from './schemas';
import type { Signer } from './schemas';

// -----------------------------------------------------------------------------
// types
// -----------------------------------------------------------------------------

export interface VerifiedRevocation {
  /** The issuer DID that revoked the credential */
  did: string;
  /** CID of the revoked credential */
  credentialCID: string;
  /** Timestamp of the revocation */
  createdAt: string;
  /** kid from the JWS header */
  signerKeyId: string;
  /** CID of the revocation artifact itself */
  revocationCID: string;
}

// -----------------------------------------------------------------------------
// signing
// -----------------------------------------------------------------------------

/**
 * Sign a revocation artifact as a JWS
 */
export const signRevocation = async (input: {
  issuerDID: string;
  credentialCID: string;
  signer: Signer;
  keyId: string;
}): Promise<{ jwsToken: string; revocationCID: string }> => {
  const kid = `${input.issuerDID}#${input.keyId}`;
  const now = new Date().toISOString().replace(/\d{3}Z$/, '000Z');

  const payload = {
    type: 'revocation' as const,
    did: input.issuerDID,
    credentialCID: input.credentialCID,
    createdAt: now,
  };

  const encoded = await dagCborCanonicalEncode(payload);
  const revocationCID = encoded.cid.toString();

  const jwsToken = await createJws({
    header: { alg: 'EdDSA', typ: 'did:dfos:revocation', kid, cid: revocationCID },
    payload: payload as unknown as Record<string, unknown>,
    sign: input.signer,
  });

  return { jwsToken, revocationCID };
};

// -----------------------------------------------------------------------------
// verification
// -----------------------------------------------------------------------------

/**
 * Verify a revocation JWS — signature, CID, payload schema, signer match
 */
export const verifyRevocation = async (input: {
  jwsToken: string;
  resolveKey: (kid: string) => Promise<Uint8Array>;
}): Promise<VerifiedRevocation> => {
  const decoded = decodeJwsUnsafe(input.jwsToken);
  if (!decoded) throw new Error('failed to decode revocation JWS');

  // parse payload
  const result = RevocationPayload.safeParse(decoded.payload);
  if (!result.success) {
    const messages = result.error.issues.map((e) => e.message).join(', ');
    throw new Error(`invalid revocation payload: ${messages}`);
  }
  const payload = result.data;

  // verify typ
  if (decoded.header.typ !== 'did:dfos:revocation') {
    throw new Error(`invalid revocation typ: ${decoded.header.typ}`);
  }

  // verify kid DID matches payload did (only the issuer can revoke)
  const kid = decoded.header.kid;
  const hashIdx = kid.indexOf('#');
  if (hashIdx < 0) throw new Error('revocation kid must be a DID URL');
  const kidDid = kid.substring(0, hashIdx);
  if (kidDid !== payload.did) {
    throw new Error('revocation kid DID does not match payload did');
  }

  // verify signature
  const publicKey = await input.resolveKey(kid);
  try {
    verifyJws({ token: input.jwsToken, publicKey });
  } catch {
    throw new Error('invalid revocation signature');
  }

  // verify CID
  const encoded = await dagCborCanonicalEncode(payload);
  const revocationCID = encoded.cid.toString();
  if (!decoded.header.cid) throw new Error('missing cid in revocation header');
  if (decoded.header.cid !== revocationCID) throw new Error('revocation cid mismatch');

  return {
    did: payload.did,
    credentialCID: payload.credentialCID,
    createdAt: payload.createdAt,
    signerKeyId: kid,
    revocationCID,
  };
};

File: packages/dfos-protocol-go/revocation.go
package dfos

import (
	"crypto/ed25519"
	"fmt"
	"strings"
)

// SignRevocation signs a credential revocation.
func SignRevocation(did, credentialCID, kid string, privateKey ed25519.PrivateKey) (jwsToken string, revocationCID string, err error) {
	now := protocolTimestamp()

	payload := map[string]any{
		"version":       int64(1),
		"type":          "revocation",
		"did":           did,
		"credentialCID": credentialCID,
		"createdAt":     now.Format("2006-01-02T15:04:05.000Z"),
	}

	_, _, cidStr, err := DagCborCID(payload)
	if err != nil {
		return "", "", err
	}

	header := JWSHeader{
		Alg: "EdDSA",
		Typ: "did:dfos:revocation",
		Kid: kid,
		CID: cidStr,
	}

	jwsToken, err = CreateJWS(header, payload, privateKey)
	if err != nil {
		return "", "", err
	}

	return jwsToken, cidStr, nil
}

// VerifiedRevocationResult is the result of revocation verification.
type VerifiedRevocationResult struct {
	// The issuer DID that revoked the credential
	DID string
	// CID of the revoked credential
	CredentialCID string
	// Timestamp of the revocation
	CreatedAt string
	// kid from the JWS header
	SignerKeyId string
	// CID of the revocation artifact itself
	RevocationCID string
}

// VerifyRevocation verifies a revocation JWS — signature, CID, payload
// schema, signer match.
func VerifyRevocation(jwsToken string, resolveKey KeyResolver) (*VerifiedRevocationResult, error) {
	return verifyRevocationCore(jwsToken, resolveKey)
}

// VerifyRevocationAt is an alias for VerifyRevocation. Revocations have no
// temporal validity window (they are permanent), so there is no custom time
// parameter. Provided for API consistency with other verify functions.
func VerifyRevocationAt(jwsToken string, resolveKey KeyResolver) (*VerifiedRevocationResult, error) {
	return verifyRevocationCore(jwsToken, resolveKey)
}

func verifyRevocationCore(jwsToken string, resolveKey KeyResolver) (*VerifiedRevocationResult, error) {
	header, payload, err := DecodeJWSUnsafe(jwsToken)
	if err != nil {
		return nil, fmt.Errorf("failed to decode revocation JWS")
	}

	// verify typ
	if header.Typ != "did:dfos:revocation" {
		return nil, fmt.Errorf("invalid revocation typ: %s", header.Typ)
	}

	// validate payload
	version, _ := payload["version"].(int64)
	if version != 1 {
		return nil, fmt.Errorf("invalid revocation payload: invalid or missing version")
	}
	if payloadString(payload, "type") != "revocation" {
		return nil, fmt.Errorf("invalid revocation payload: wrong type")
	}
	did := payloadString(payload, "did")
	if did == "" {
		return nil, fmt.Errorf("invalid revocation payload: missing did")
	}
	credentialCID := payloadString(payload, "credentialCID")
	if credentialCID == "" {
		return nil, fmt.Errorf("invalid revocation payload: missing credentialCID")
	}
	createdAt := payloadString(payload, "createdAt")
	if err := validateCreatedAt(createdAt); err != nil {
		return nil, fmt.Errorf("invalid revocation payload: %w", err)
	}

	// verify kid DID matches payload did (only the issuer can revoke)
	kid := header.Kid
	hashIdx := strings.Index(kid, "#")
	if hashIdx < 0 {
		return nil, fmt.Errorf("revocation kid must be a DID URL")
	}
	kidDid := kid[:hashIdx]
	if kidDid != did {
		return nil, fmt.Errorf("revocation kid DID does not match payload did")
	}

	// verify signature
	publicKey, err := resolveKey(kid)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve revocation key: %w", err)
	}
	if _, _, err := VerifyJWS(jwsToken, publicKey); err != nil {
		return nil, fmt.Errorf("invalid revocation signature")
	}

	// verify CID
	_, _, revocationCID, err := DagCborCID(payload)
	if err != nil {
		return nil, fmt.Errorf("failed to derive revocation CID: %w", err)
	}
	if header.CID == "" {
		return nil, fmt.Errorf("missing cid in revocation header")
	}
	if header.CID != revocationCID {
		return nil, fmt.Errorf("revocation cid mismatch")
	}

	return &VerifiedRevocationResult{
		DID:           did,
		CredentialCID: credentialCID,
		CreatedAt:     createdAt,
		SignerKeyId:   kid,
		RevocationCID: revocationCID,
	}, nil
}
